### PR TITLE
 feat: new component tooltip with button knockout for escher map.

### PR DESCRIPTION
**Pull request by AlbaLopez**

### CODE REVIEW

File: src/pathwayvis/components/map/map.component.ts [hunk 1]
```diff
@@ -1,8 +1,8 @@
 import * as escher from '@dd-decaf/escher';
 import * as d3 from 'd3';
-import {event as currentEvent} from 'd3';
+import { event as currentEvent } from 'd3';
 import * as _ from 'lodash';
-
+import * as tinier from 'tinier';
```
Comment (matyasfodor) [inline src/pathwayvis/components/map/map.component.ts]: pls keep the blank line between 3rd party imports and local imports

File: src/pathwayvis/components/tooltip/newTooltip.ts [hunk 2]
```diff
@@ -0,0 +1,133 @@
+import * as tinier from 'tinier';
+import * as _ from 'underscore';
+
+function decompartmentalize(id) {
```
Comment (matyasfodor) [inline src/pathwayvis/components/tooltip/newTooltip.ts]: There's no way to access these helpers form escher?

Comment (matyasfodor) [reply src/pathwayvis/components/tooltip/newTooltip.ts]: Is this file copied over from escher? There was no way to extend the code coming form escher? If no, could you please highlight where you extended / modified the original code?

File: src/pathwayvis/components/map/map.component.ts [hunk 3]
```diff
@@ -429,41 +453,40 @@ class MapComponentCtrl {
...
-        if (this.contextElement) {
-          this._renderContextMenu(contextMenu);
-          (<MouseEvent> currentEvent).preventDefault();
-        }
-      });
 
-    d3.select(document).on('click', () => {
-      contextMenu.style('visibility', 'hidden');
+  private _getContext(): void {
+    let data = JSON.parse(d3.select('div#tooltip-container').select('#knockoutbutton').attr("data").
```
Comment (matyasfodor) [inline src/pathwayvis/components/map/map.component.ts]: Why `let`?

Comment (matyasfodor) [reply src/pathwayvis/components/map/map.component.ts]: I think it's better if you change the keys/values on the js object:

const {biggId, ...restData} = JSON.parse(d3.select('div#tooltip-container').select('#knockoutbutton').attr("data"))
const data = Object.assign({bigg_id: biggId}, ...restData)

File: src/pathwayvis/components/map/map.component.ts [hunk 4]
```diff
@@ -429,41 +453,40 @@ class MapComponentCtrl {
...
+    replace("biggId", "bigg_id"));
+    this.contextElement = data;
+    this.contextActions = this.actions.getList({
+      type: 'map:reaction',
+      shared: this._mapOptions.getMapData(),
+      element: this.contextElement,
     });
+   d3.select('div#tooltip-container').select('#knockoutbutton').text(this.contextActions[0].label);
```
Comment (matyasfodor) [inline src/pathwayvis/components/map/map.component.ts]: You could extract `d3.select('div#tooltip-container')` from these two lines.

File: src/pathwayvis/components/map/map.component.ts [hunk 5]
```diff
@@ -429,41 +453,40 @@ class MapComponentCtrl {
...
 
-        if (this.contextElement) {
-          this._renderContextMenu(contextMenu);
-          (<MouseEvent> currentEvent).preventDefault();
-        }
-      });
 
-    d3.select(document).on('click', () => {
-      contextMenu.style('visibility', 'hidden');
+  private _getContext(): void {
```
Comment (matyasfodor) [inline src/pathwayvis/components/map/map.component.ts]: Could you briefly describe the purpose of this function?

File: src/pathwayvis/components/map/views/map.component.html [hunk 6]
```diff
@@ -1,7 +1,7 @@
 <md-whiteframe class="map-context-menu md-whiteframe-4dp" layout="column" layout-align="center center" ng-cloak style="visibility: hidden">
   <md-menu-content class="context-menu">
     <md-menu-item ng-repeat="action in ctrl.contextActions">
-      <md-button ng-click="ctrl.processActionClick(action, ctrl.contextElement)">{{action.label}}</md-button>
+      <md-button ng-click="ctrl.processActionClick(action, ctrl.contextElement)">{{JSON.stringfy(ctrl.contextActions.shared)}}</md-button>
```
Comment (matyasfodor) [inline src/pathwayvis/components/map/views/map.component.html]: It'd be nice not to add logic to templates. Could you move this to the controller?

File: src/pathwayvis/components/map/views/map.component.scss [hunk 7]
```diff
@@ -57,4 +57,56 @@ pv-map {
 
 .context-menu {
     width: 100%
-}
\ No newline at end of file
+}
+.container-style {
```
Comment (matyasfodor) [inline src/pathwayvis/components/map/views/map.component.scss]: All css should go in the element's selector (`pv-map` in this case). I know you didn't introduce this change, but it'd be nice if you could fix it. Could you please check if it's still working if you move this part of the code within 
```
pv-map {
  ...
}
```
?

File: src/pathwayvis/components/reactionpanel/knockouts/knockouts.component.ts [hunk 8]
```diff
@@ -31,6 +31,7 @@ class KnockoutsController {
   public removeReactionSelectedItem(item) {
     if (!item) return;
     const doKnockoutAction = this._actions.getAction('reaction:knockout:do');
+    if (item.id === this._mapOptions.getObjectiveReaction()) this._mapOptions.setObjectiveReaction('');
```
Comment (matyasfodor) [inline src/pathwayvis/components/reactionpanel/knockouts/knockouts.component.ts]: Please always create a block after if statements

File: src/pathwayvis/components/reactionpanel/objectives/objectives.component.ts [hunk 9]
```diff
@@ -0,0 +1,81 @@
+import * as angular from "angular";
+
+import { MapOptionService } from "../../../services/mapoption.service";
+import { ActionsService } from "../../../services/actions/actions.service";
+import { access } from "fs";
```
Comment (matyasfodor) [inline src/pathwayvis/components/reactionpanel/objectives/objectives.component.ts]: Why would you need the file system?

**Review by matyasfodor — CHANGES_REQUESTED** [review]

File: src/pathwayvis/components/map/map.component.ts [hunk 10]
```diff
@@ -473,18 +473,21 @@ class MapComponentCtrl {
...
 
-
+ /**
+  * Get context menu and fetches list of actions for selected map element
+  */
   private _getContext(): void {
-    let data = JSON.parse(d3.select('div#tooltip-container').select('#knockoutbutton').attr("data").
-    replace("biggId", "bigg_id"));
+    const tooltipContainer = d3.select('div#tooltip-container');
+    const {biggId, ...restData} = JSON.parse(d3.select('div#tooltip-container').select('#knockoutbutton').attr("data"));
```
Comment (matyasfodor) [inline src/pathwayvis/components/map/map.component.ts]: Here you can refer the variable you created in the line above.

File: src/pathwayvis/components/map/map.component.ts [hunk 11]
```diff
@@ -473,18 +473,21 @@ class MapComponentCtrl {
...
+    const data = Object.assign({bigg_id: biggId}, ...restData);
```
Comment (matyasfodor) [inline src/pathwayvis/components/map/map.component.ts]: I just realized, there's no point in creating data since its only use is in the assignment in the line below.
So you could just go like `this.contextElement = Object.assign({bigg_id: biggId}, ...restData);`

File: src/pathwayvis/components/map/map.component.ts (hunk 5 above)
Comment (matyasfodor) [reply src/pathwayvis/components/map/map.component.ts]: I didn't mean a comment, but it's also nice to have. although I still don't completely understand the purpose of this function in detail.

File: src/pathwayvis/components/map/views/map.component.scss [hunk 12]
```diff
@@ -53,60 +53,59 @@ pv-map {
...
+        top: 4px;
+        right: 4px;
+        color: #d27066;
+        background-color: #ffeded;
+        border-radius: 2px;
+        font-size: 14px;
+        text-align: right;
+        padding: 0px 5px;
+      }
+}
```
Comment (matyasfodor) [inline src/pathwayvis/components/map/views/map.component.scss]: Please add a trailing blankline

File: src/pathwayvis/components/map/map.component.ts [hunk 13]
```diff
@@ -143,12 +143,12 @@ class MapComponentCtrl {
             return reaction.id.startsWith(str);
           });
         })
-        .map((reaction) => {
-          return Object.assign({}, reaction, {
-            bigg_id: reaction.id,
-            metabolites: reaction.metabolites,
+          .map((reaction) => {
```
Comment (matyasfodor) [inline src/pathwayvis/components/map/map.component.ts]: In this case, you can go with a body-less arrow function:

    (reaction) => Object.assign({}, reaction, {
       bigg_id: reaction.id,
       metabolites: reaction.metabolites,
     }))

File: src/pathwayvis/components/map/map.component.ts [hunk 14]
```diff
@@ -330,16 +329,38 @@ class MapComponentCtrl {
   /**
    * Callback function for clicked action button in context menu
    */
-  public processActionClick(action, data) {
-
+  public processActionClick(action, data, element) {
     if (action.type === 'reaction:link') {
       this.$window.open(`http://bigg.ucsd.edu/universal/reactions/${data.bigg_id}`);
+    } else if (action.type === "reaction:objective:do") {
```
Comment (matyasfodor) [inline src/pathwayvis/components/map/map.component.ts]: There's a bit of a code duplication here. I think this would be better

```
} else if (action.type.startsWith('reaction:objective')) {
  this.shared.async(this._mapOptions.actionHandler(action, { id: data.bigg_id }).then((response) => {
    if (action.type.endsWith('do')) {
        this._mapOptions.setCurrentGrowthRate(parseFloat(response['growth-rate']));
        this._mapOptions.setReactionData(response.fluxes);
        this._mapOptions.setRemovedReactions(response['removed-reactions']);
        this._mapOptions.setObjectiveReaction(data.bigg_id);
        this._getContext();
    } else {
        this._mapOptions.setObjectiveReaction("");
        this._loadMap(this._mapOptions.getType(), this._mapOptions.getDataObject().selected, this._mapOptions.getSelectedId());
    }
  }, (error) => {
    this.toastService.showErrorToast('Oops! Sorry, there was a problem.');
  }), 'Objective');
}
```

Looking at this now, the third case (`else`) could be also merged into this logic.

File: src/pathwayvis/components/map/map.component.ts [hunk 15]
```diff
@@ -426,43 +449,46 @@ class MapComponentCtrl {
...
-          shared: this._mapOptions.getMapData(),
-          element: this.contextElement,
+    const tooltipContainer = d3.select('div#tooltip-container');
+    d3.selectAll('text.reaction-label').on('mouseenter', (d) => {
+      if (tooltipContainer.select('#knockoutbutton')) {
+        this._getContext();
+        tooltipContainer.select('#knockoutbutton').on('click', () => {
+          this.processActionClick(this.contextActions[0], this.contextElement, this);
+          if (this.contextElement) {
+            this.$scope.$apply();
```
Comment (matyasfodor) [inline src/pathwayvis/components/map/map.component.ts]: why is this required? This is usually a bad practice, so it should be avoided if possible

File: src/pathwayvis/components/reactionpanel/knockouts/knockouts.component.ts [hunk 16]
```diff
@@ -31,6 +31,9 @@ class KnockoutsController {
   public removeReactionSelectedItem(item) {
     if (!item) return;
     const doKnockoutAction = this._actions.getAction('reaction:knockout:do');
+    if (item.id === this._mapOptions.getObjectiveReaction()) {
+      this._mapOptions.setObjectiveReaction('');
```
Comment (matyasfodor) [inline src/pathwayvis/components/reactionpanel/knockouts/knockouts.component.ts]: Is there any way to do this somewhere else? The knockout component has nothing to do with Objective

File: src/pathwayvis/components/reactionpanel/objectives/objectives.component.ts [hunk 17]
```diff
@@ -0,0 +1,80 @@
+import * as angular from "angular";
+
+import { MapOptionService } from "../../../services/mapoption.service";
+import { ActionsService } from "../../../services/actions/actions.service";
+
+const DecafBiggProxy = 'https://api-staging.dd-decaf.eu/bigg/';
```
Comment (matyasfodor) [inline src/pathwayvis/components/reactionpanel/objectives/objectives.component.ts]: Please do not hardcode this. bigg is already being fetched in added reactions, I guess it's accessed through a provider.

File: src/pathwayvis/components/reactionpanel/objectives/objectives.component.ts [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+
+class ObjectivesController {
+  public modelReactions = [];
```
Comment (matyasfodor) [inline src/pathwayvis/components/reactionpanel/objectives/objectives.component.ts]: Perhaps you should give this one a type

File: src/pathwayvis/components/reactionpanel/objectives/objectives.component.ts [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+    actions: ActionsService) {
+    this._mapOptions = mapOptions;
+    this._actions = actions;
+
+    mapOptions.reactionsObservable.subscribe((reactions) => {
+      this.modelReactions = reactions;
+    });
+
+    mapOptions.objectiveReactionObservable.subscribe((reaction) => {
+      this.objectiveReaction = [];
```
Comment (matyasfodor) [inline src/pathwayvis/components/reactionpanel/objectives/objectives.component.ts]: Why do you set it to empty list? You could just use a simple element instead of an array, and set it to `null` if it is removed. Or are you planning to support multiple objectives?

File: src/pathwayvis/components/reactionpanel/objectives/objectives.component.ts [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+
+  public onUndoClick(selectedReaction: string): void {
+    const undoObjectiveAction = this._actions.getAction('reaction:objective:undo');
+    this._mapOptions.setObjectiveReaction('');
+    this.objectiveReaction = [];
+    this._mapOptions.actionHandler(undoObjectiveAction, { id: selectedReaction })
+    .then(this.updateMapData.bind(this));
+  }
+
+  public objectiveDisplay(item) {
```
Comment (matyasfodor) [inline src/pathwayvis/components/reactionpanel/objectives/objectives.component.ts]: Instead of doing this, you could extend `rp-panel-item` to have `item-display` as an optional attribute. If it's not passed, it could fall back to this dummy function.

File: src/pathwayvis/components/reactionpanel/objectives/objectives.component.ts [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+    this._mapOptions.actionHandler(doObjectiveAction, { id: item.id })
+    .then(this.updateMapData.bind(this));
+  }
+
+  public onUndoClick(selectedReaction: string): void {
+    const undoObjectiveAction = this._actions.getAction('reaction:objective:undo');
+    this._mapOptions.setObjectiveReaction('');
+    this.objectiveReaction = [];
+    this._mapOptions.actionHandler(undoObjectiveAction, { id: selectedReaction })
+    .then(this.updateMapData.bind(this));
```
Comment (matyasfodor) [inline src/pathwayvis/components/reactionpanel/objectives/objectives.component.ts]: indentation is off

File: src/pathwayvis/components/reactionpanel/objectives/objectives.component.ts [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+  }
+
+  public objectiveDisplay(item) {
+    return item;
+  }
+
+  private updateMapData(response) {
+    this._mapOptions.setCurrentGrowthRate(parseFloat(response['growth-rate']));
```
Comment (matyasfodor) [inline src/pathwayvis/components/reactionpanel/objectives/objectives.component.ts]: I see this bit of code repeating everywhere in the code. This might be challenging, but could you check if it's feasible the do this only in one place?

File: src/pathwayvis/components/reactionpanel/objectives/objectives.component.ts [hunk 23]
```diff
@@ -0,0 +1,80 @@
...
+    this._mapOptions.setReactionData(response.fluxes);
+    this._mapOptions.setDataModel(response.model);
+    this._mapOptions.setRemovedReactions(response['removed-reactions']);
+  }
+}
+
+export const ObjectivesComponent = {
+  controller: ObjectivesController,
+  template: `
+  <rp-panel-item
```
Comment (matyasfodor) [inline src/pathwayvis/components/reactionpanel/objectives/objectives.component.ts]: What do you think about this element? Is it useful? Was it easy to use?

File: src/pathwayvis/services/actions/actions.service.ts [hunk 24]
```diff
@@ -43,8 +43,8 @@ export class ActionsService {
    * @return {types.Action} Action
    */
   public getAction(type: string): types.Action {
-    actionsList = actionsList.filter((action: types.Action) => action.type === type);
-    return actionsList[0] || null;
+    let actionSelected = actionsList.filter((action: types.Action) => action.type === type);
```
Comment (matyasfodor) [inline src/pathwayvis/services/actions/actions.service.ts]: I guess this could be const.

File: src/pathwayvis/services/actions/actions.service.ts [hunk 25]
```diff
@@ -87,6 +87,36 @@ class UndoKnockout extends Knockout {
   }
 }
 
+@registerAction
+// tslint:disable-next-line
+class SetObjective extends ReactionSetAsObjectiveAction implements Action {
+  public label = 'Set as objective';
+  public type: string = 'reaction:objective:do';
+  public shared: types.MapData;
+
```
Comment (matyasfodor) [inline src/pathwayvis/services/actions/actions.service.ts:95]: Only one blank line please

File: src/pathwayvis/services/actions/base.ts [hunk 26]
```diff
@@ -36,3 +36,21 @@ export abstract class ReactionAction extends Action {
...
 }
+/**
+ * Abstract class for Reaction actions with predefined context type
+ */
+export abstract class ReactionSetAsObjectiveAction extends Action {
+  public shared: MapData;
+
+  public callback(ws: WSService, $timeout: angular.ITimeoutService, shared: SharedService): any {
+
+    const data = {
```
Comment (matyasfodor) [inline src/pathwayvis/services/actions/base.ts]: I guess you could just inline this variable

File: src/pathwayvis/services/mapoption.service.ts [hunk 27]
```diff
@@ -306,6 +321,10 @@ export class MapOptionService {
       }
     } else if (action.type === 'reaction:update') {
       if (reaction) shared.addedReactions.push(reaction);
+    } else if (action.type === 'reaction:objective:do') {
+      if (id) shared.objectiveReaction = id;
```
Comment (matyasfodor) [inline src/pathwayvis/services/mapoption.service.ts]: Please use a block here

File: src/pathwayvis/services/mapoption.service.ts [hunk 28]
```diff
@@ -306,6 +321,10 @@ export class MapOptionService {
...
+    } else if (action.type === 'reaction:objective:undo') {
+      shared.objectiveReaction = '';
```
Comment (matyasfodor) [inline src/pathwayvis/services/mapoption.service.ts]: I think it's better to use `null`

**Review by matyasfodor — CHANGES_REQUESTED** [review]

File: src/pathwayvis/components/map/map.component.ts [hunk 29]
```diff
@@ -330,16 +327,35 @@ class MapComponentCtrl {
...
+        } else {
+            this._mapOptions.setObjectiveReaction("");
+            this._loadMap(this._mapOptions.getType(), this._mapOptions.getDataObject().selected, this._mapOptions.getSelectedId());
+        }
+      }, (error) => {
+        this.toastService.showErrorToast('Oops! Sorry, there was a problem.');
+      }), 'Objective');
     } else {
-      this._mapOptions.actionHandler(action, { id: data.bigg_id }).then((response) => {
+      this.shared.async(this._mapOptions.actionHandler(action, { id: data.bigg_id }).then((response) => {
```
Comment (matyasfodor) [inline src/pathwayvis/components/map/map.component.ts]: Could you also merge this one to here? https://github.com/DD-DeCaF/metabolica-ui-map/pull/141/files#diff-62c06d0a48047fe88322f20be6e856e2R334

File: src/pathwayvis/components/map/map.component.ts [hunk 30]
```diff
@@ -386,16 +403,17 @@ class MapComponentCtrl {
...
-      reactionsToHighlight.forEach((reactionId) => {
-        this._builder.map.bigg_index
-          .getAll(reactionId)
-          .forEach(({reaction_id}) => {
-            d3.select(`#r${reaction_id}`)
-              .style("filter", "url(#escher-glow-filter)");
-          });
-      });
+      if (model.notes.changes) {
+        const reactionsToHighlight = model.notes.changes.measured.reactions.map((reaction) => reaction.id);
```
Comment (matyasfodor) [inline src/pathwayvis/components/map/map.component.ts]: Looking at this `reactionsToHighlight` could be inlined, just add a few linebreaks. Thank you!

File: src/pathwayvis/components/map/views/map.component.scss [hunk 31]
```diff
@@ -42,19 +45,75 @@ pv-map {
...
+        top: 4px;
+        right: 4px;
+        color: #d27066;
+        background-color: #ffeded;
+        border-radius: 2px;
+        font-size: 14px;
+        text-align: right;
+        padding: 0px 5px;
+      }
 }
```
Comment (matyasfodor) [inline src/pathwayvis/components/map/views/map.component.scss]: I meant adding a blank line at the end of the file

File: src/pathwayvis/components/reactionpanel/knockouts/knockouts.component.ts [hunk 32]
```diff
@@ -31,6 +31,9 @@ class KnockoutsController {
   public removeReactionSelectedItem(item) {
     if (!item) return;
     const doKnockoutAction = this._actions.getAction('reaction:knockout:do');
+    if (item.id === this._mapOptions.getObjectiveReaction()) {
```
Comment (matyasfodor) [inline src/pathwayvis/components/reactionpanel/knockouts/knockouts.component.ts]: So there's no way of doing this elsewhere?

**Review by matyasfodor — CHANGES_REQUESTED** [review]

File: src/pathwayvis/components/map/map.component.ts [hunk 33]
```diff
@@ -330,19 +327,43 @@ class MapComponentCtrl {
...
+            this._mapOptions.setObjectiveReaction(data.bigg_id);
+            this._getContext();
+        }
+      }, (error) => {
+        this.toastService.showErrorToast('Oops! Sorry, there was a problem.');
+      }), 'Objective');
     } else {
-      this._mapOptions.actionHandler(action, { id: data.bigg_id }).then((response) => {
+      this.shared.async(this._mapOptions.actionHandler(action, { id: data.bigg_id }).then((response) => {
         this._mapOptions.setCurrentGrowthRate(parseFloat(response['growth-rate']));
```
Comment (matyasfodor) [inline src/pathwayvis/components/map/map.component.ts]: This one could also use `mapoptions.updateMapData`?

File: src/pathwayvis/components/map/map.component.ts [hunk 34]
```diff
@@ -330,19 +327,43 @@ class MapComponentCtrl {
...
+      }, (error) => {
+        this.toastService.showErrorToast('Oops! Sorry, there was a problem.');
+      }), 'Experiment');
     }
   }
-
+  /**
+   * Set map options
+   */
+  private _setMapOption(){
```
Comment (matyasfodor) [inline src/pathwayvis/components/map/map.component.ts]: Why is this empty here?

File: src/pathwayvis/components/reactionpanel/addedReactions/addedReactions.component.ts [hunk 35]
```diff
@@ -60,10 +60,7 @@ class AddedReactionsController {
   }
 
   private updateMapData(response) {
```
Comment (matyasfodor) [inline src/pathwayvis/components/reactionpanel/addedReactions/addedReactions.component.ts]: Why don't you call `this._mapOptions.updateMapData(response);` Then this method would be unnecessary.

**Review by matyasfodor — CHANGES_REQUESTED** [review]

File: src/pathwayvis/components/reactionpanel/objectives/objectives.component.ts [hunk 36]
```diff
@@ -64,7 +60,7 @@ export const ObjectivesComponent = {
     on-item-select="$ctrl.setAsObjectiveSelectedItem(item)"
     query-search="$ctrl.queryModelReactions(query)"
     placeholder="'Enter the reaction you want to remove'"
-    missing-items="'No reaction set as objective'"
+    missing-items="'Nothing as objective'"
```
Comment (matyasfodor) [inline src/pathwayvis/components/reactionpanel/objectives/objectives.component.ts]: I think this doesn't make sense

File: src/pathwayvis/services/mapoption.service.ts [hunk 37]
```diff
@@ -416,7 +416,7 @@ export class MapOptionService {
   }
 
   public updateMapData(data) {
-    this.setCurrentGrowthRate(parseFloat(data['growth-rate']));
+    this.setCurrentGrowthRate(data['growth-rate']);
```
Comment (matyasfodor) [inline src/pathwayvis/services/mapoption.service.ts]: So there's no need to parse it? I think it's required

File: src/pathwayvis/components/map/views/map.component.scss [hunk 38]
```diff
@@ -53,8 +53,56 @@ pv-map {
...
+        background-color: #ffeded;
+        border-radius: 2px;
+        font-size: 14px;
+        text-align: right;
+        padding: 0px 5px;
+      }
 
-.context-menu {
-    width: 100%
 }
```
Comment (matyasfodor) [inline src/pathwayvis/components/map/views/map.component.scss]: There's still _no_ blank line at the end pf the file

File: src/pathwayvis/components/map/map.component.ts [hunk 39]
```diff
@@ -330,16 +327,27 @@ class MapComponentCtrl {
...
     } else {
-      this._mapOptions.actionHandler(action, { id: data.bigg_id }).then((response) => {
-        this._mapOptions.setCurrentGrowthRate(parseFloat(response['growth-rate']));
-        this._mapOptions.setReactionData(response.fluxes);
-        this._mapOptions.setRemovedReactions(response['removed-reactions']);
-      });
+      this.shared.async(this._mapOptions.actionHandler(action, { id: data.bigg_id }).then((response) => {
+        this._mapOptions.updateMapData(response);
+        if (action.type.startsWith('reaction:objective')) {
+          if (action.type.endsWith('undo')) {
```
Comment (matyasfodor) [inline src/pathwayvis/components/map/map.component.ts]: We have a little bit too much of nested if statements here.This one could be simplified:

    this._mapOptions.setObjectiveReaction(action.type.endsWith('do') ? data.bigg_id : null)

File: src/pathwayvis/components/map/map.component.ts [hunk 40]
```diff
@@ -330,16 +327,27 @@ class MapComponentCtrl {
...
+            this._mapOptions.setObjectiveReaction(null);
+          } else if (action.type.endsWith('do')) {
+            this._mapOptions.setObjectiveReaction(data.bigg_id);
+          }
+        } else {
```
Comment (matyasfodor) [inline src/pathwayvis/components/map/map.component.ts]: Let's merge the two cases. Instead of

        } else {
          if (data.bigg_id === this._mapOptions.getObjectiveReaction()) {
            this._mapOptions.setObjectiveReaction(null);
          }
        }

You could go like:

        } else if (data.bigg_id === this._mapOptions.getObjectiveReaction()) {
          this._mapOptions.setObjectiveReaction(null);
        }

File: src/pathwayvis/components/reactionpanel/panelitem/panelitem.component.ts [hunk 41]
```diff
@@ -13,11 +13,16 @@ class PanelItemCtrl {
   public onRemoveItem: (item: any) => void;
   public itemDisplay: (item: any) => string;
   public idProperty: string;
-
+  public objectiveReaction: string = null;
```
Comment (matyasfodor) [inline src/pathwayvis/components/reactionpanel/panelitem/panelitem.component.ts]: What's the purpose of this? It's not added to the bindings. Also, I don't really like the name since it's too specific. If you look at `panelitem`, it's very generic, you could use it to build a horse race betting site! I'd either turn `items` => `item` and allow it to accept a single element and an array as well (and dispatch in the component). Or you could have `items` and `item` next to each other. if `items` is missing, `item` would be displayed.

File: src/pathwayvis/services/actions/actions.service.ts [hunk 42]
```diff
@@ -87,6 +87,35 @@ class UndoKnockout extends Knockout {
...
+  }
+}
+
+@registerAction
+// tslint:disable-next-line
+class UndoSetObjective extends ReactionAction {
+  public label = 'Undo set as objective';
+  public type: string = 'reaction:objective:undo';
+
+  public canDisplay(context) {
```
Comment (matyasfodor) [inline src/pathwayvis/services/actions/actions.service.ts]: What if we'd turn this into a multi-line logical expression?

    public canDisplay({shared, element}) {
      return shared.objectiveReaction &&
        shared.objectiveReaction === element.bigg_id &&
        context.type === 'map:reaction';
    }

File: src/pathwayvis/services/mapoption.service.ts [hunk 43]
```diff
@@ -291,21 +306,32 @@ export class MapOptionService {
     this.selectedCardId = id;
   }
 
-  public actionHandler(
+   public actionHandler(
```
Comment (matyasfodor) [inline src/pathwayvis/services/mapoption.service.ts]: I guess it's a mistake. Or the indentation was off?

File: src/pathwayvis/services/mapoption.service.ts [hunk 44]
```diff
@@ -388,4 +414,11 @@ export class MapOptionService {
   public getSharedPathway(): any {
     return this.getMapData().pathwayData;
   }
+
+  public updateMapData(data) {
+    this.setCurrentGrowthRate(parseFloat(data['growth-rate']));
+    this.setReactionData(data.fluxes);
+    this.setDataModel(data.model);
```
Comment (matyasfodor) [inline src/pathwayvis/services/mapoption.service.ts:421]: Is this working after all?

Comment (AlbaLopez) [reply src/pathwayvis/services/mapoption.service.ts:421]: yes!

File: src/pathwayvis/services/mapoption.service.ts (hunk 43 above)
Comment (AlbaLopez) [reply src/pathwayvis/services/mapoption.service.ts]: mistake

File: src/pathwayvis/components/map/map.component.ts [hunk 45]
```diff
@@ -330,16 +327,21 @@ class MapComponentCtrl {
   /**
    * Callback function for clicked action button in context menu
    */
-  public processActionClick(action, data) {
-
+  public processActionClick(action, data, element) {
```
Comment (matyasfodor) [inline src/pathwayvis/components/map/map.component.ts]: I guess you won't need `element` here eventually

File: src/pathwayvis/components/map/map.component.ts [hunk 46]
```diff
@@ -1,21 +1,21 @@
...
-
 import * as types from '../../types';
-
 import './views/map.component.scss';
 import * as template from './views/map.component.html';
 import { ToastService } from "../../services/toastservice";
 import { MapOptionService } from "../../services/mapoption.service";
 import { ObjectType } from "../../types";
 import * as utils from "../../utils";
+import * as escherTooltip from "../tooltip/escherTooltip";
```
Comment (matyasfodor) [inline src/pathwayvis/components/map/map.component.ts]: why don't you import `{ Tooltip }`? And then you could just pass `Tooltip` here https://github.com/DD-DeCaF/metabolica-ui-map/pull/141/files#diff-62c06d0a48047fe88322f20be6e856e2R364

File: src/pathwayvis/components/map/map.component.ts [hunk 47]
```diff
@@ -426,43 +431,39 @@ class MapComponentCtrl {
   }
 
   /**
-   * Loads context menu and fetches list of actions for selected map element
+   * Loads context menu with _getContext method when you hover a reaction.
```
Comment (matyasfodor) [inline src/pathwayvis/components/map/map.component.ts]: ~~hover~~ hover over

File: src/pathwayvis/components/reactionpanel/panelitem/panelitem.component.html [hunk 48]
```diff
@@ -17,13 +17,20 @@
       </md-item-template>
     </md-autocomplete>
     <md-list>
-      <md-list-item ng-hide="$ctrl.items.length > 0">{{ $ctrl.missingItems }}</md-list-item>
-      <md-list-item ng-repeat="item in $ctrl.items">
-        <span class="md-caption">{{ $ctrl.itemDisplay({item: item}) }}</span>
+      <md-list-item ng-hide="$ctrl.item || $ctrl.items.length > 0">{{ $ctrl.missingItems }}</md-list-item>
+      <md-list-item ng-if="$ctrl.item">
+        <span class="md-caption">{{ $ctrl.item }}</span>
```
Comment (matyasfodor) [inline src/pathwayvis/components/reactionpanel/panelitem/panelitem.component.html]: I think remove is missing here

**Review by matyasfodor — APPROVED** [review]
